src/core/tsi/s2a/record_protocol/s2a_crypter.cc

```diff
@@ -598,6 +602,231 @@ grpc_status_code s2a_encrypt(s2a_crypter* crypter, uint8_t* plaintext,
   return s2a_write_tls13_record(crypter, SSL3_RT_APPLICATION_DATA,
-                                &plaintext_vec, 1, record_vec, record_size,
-                                error_details);
+                                &plaintext_vec, /** plaintext_vec length **/ 1,
+                                record_vec, record_size, error_details);
+}
+
+grpc_status_code s2a_max_plaintext_size(const s2a_crypter* crypter,
+                                        size_t record_size,
+                                        size_t* plaintext_size,
+                                        char** error_details) {
+  GPR_ASSERT(crypter != nullptr);
+  GPR_ASSERT(plaintext_size != nullptr);
+
+  /** Note that we require this method returns 1 more than the size of the
+   *  plaintext that is returned by decrypting a TLS 1.3 record of size
+   *  |record_size|; this is because |crypter| requires 1 byte of extra space
+   *  after the plaintext to decrypt the record type. **/
+  size_t tag_size;
+  grpc_status_code tag_status = s2a_tag_size(crypter, &tag_size, error_details);
+  if (tag_status != GRPC_STATUS_OK) {
+    return tag_status;
+  }
+  GPR_ASSERT(record_size >= SSL3_RT_HEADER_LENGTH + tag_size);
+  *plaintext_size = record_size - SSL3_RT_HEADER_LENGTH - tag_size;
+  return GRPC_STATUS_OK;
+}
+
+grpc_status_code s2a_max_plaintext_size(const s2a_crypter* crypter,
+                                        const iovec* protected_vec,
+                                        size_t protected_vec_size,
+                                        size_t* plaintext_size,
+                                        char** error_details) {
+  GPR_ASSERT(crypter != nullptr);
+  GPR_ASSERT(plaintext_size != nullptr);
+
+  /** Note that we require this method returns 1 more than the size of the
+   *  plaintext that is returned by decrypting the TLS 1.3 payload stored in
+   *  |protected_vec|; this is because |crypter| requires 1 byte of extra space
+   *  after the plaintext to decrypt the record type. **/
+  size_t tag_size;
+  grpc_status_code tag_status = s2a_tag_size(crypter, &tag_size, error_details);
+  if (tag_status != GRPC_STATUS_OK) {
+    return tag_status;
+  }
+  size_t total_vec_length = get_total_length(protected_vec, protected_vec_size);
+  GPR_ASSERT(total_vec_length >= tag_size);
+  *plaintext_size = total_vec_length - tag_size;
+  return GRPC_STATUS_OK;
+}
+
+s2a_decrypt_status s2a_decrypt_payload(
+    s2a_crypter* crypter, iovec& record_header, const iovec* protected_vec,
+    size_t protected_vec_size, iovec& unprotected_vec,
+    size_t* bytes_written, char** error_details) {
+  GPR_ASSERT(crypter != nullptr);
+  GPR_ASSERT(bytes_written != nullptr);
+  size_t payload_size = get_total_length(protected_vec, protected_vec_size);
+  size_t tag_size;
+  grpc_status_code tag_status = s2a_tag_size(crypter, &tag_size, error_details);
+  if (tag_status != GRPC_STATUS_OK) {
+    return INTERNAL_ERROR;
+  }
+  GPR_ASSERT(tag_size <= payload_size);
```

Comment:
Not sure if check-failing on wrong length is a good idea here (didn't find this in the docs).
Can you return something like FAILED_PRECONDITION?

Reply:
Good call - made this change.